@xen-orchestra/vmware-explorer/esxi.mjs

```diff
@@ -0,0 +1,270 @@
+        {
+          attributes: { 'xsi:type': 'TraversalSpec' },
+          name: 'traverseEntities',
+          type: 'ContainerView',
+          path: 'view',
+          skip: false,
+        },
+      ],
+    }
+
+    const propertyFilterSpec = {
+      attributes: { 'xsi:type': 'PropertyFilterSpec' },
+      propSet: properties.map(p => ({
+        attributes: { 'xsi:type': 'PropertySpec' },
+        type,
+        pathSet: [p],
+      })),
+      objectSet: [objectSpec],
+    }
+
+    result = await this.#exec('RetrievePropertiesEx', {
+      _this: propertyCollector,
+      specSet: [propertyFilterSpec],
+      options: { attributes: { type: 'RetrieveOptions' } },
+    })
+
+    const objects = {}
+    const returnObj = Array.isArray(result.returnval.objects) ? result.returnval.objects : [result.returnval.objects]
+
+    returnObj.forEach(({ obj, propSet }) => {
+      objects[obj.$value] = {}
+      propSet = Array.isArray(propSet) ? propSet : [propSet]
+      propSet.forEach(({ name, val }) => {
+        // don't care about the type for now
+        delete val.attributes
+        // a scalar value : simplify it
+        if (val.$value) {
+          objects[obj.$value][name] = val.$value
+        } else {
+          objects[obj.$value][name] = val
+        }
+      })
+    })
+    return objects
+  }
+
+  async #inspectVmdk(dataStores, currentDataStore, currentPath, filePath) {
+    let diskDataStore, diskPath
+    if (filePath.startsWith('/')) {
+      // disk is on another datastore
+      Object.keys(dataStores).forEach(dataStoreUrl => {
+        if (filePath.startsWith(dataStoreUrl)) {
+          diskDataStore = dataStores[dataStoreUrl].name
+          diskPath = filePath.substring(dataStoreUrl.length + 1)
+        }
+      })
+    } else {
+      diskDataStore = currentDataStore
+      diskPath = currentPath + '/' + filePath
+    }
+    const vmdkRes = await this.download(diskDataStore, diskPath)
+    const text = await vmdkRes.text()
+    const parsed = parseVmdk(text)
+
+    return {
+      ...parsed,
+      datastore: diskDataStore,
+      path: dirname(diskPath),
+      descriptionLabel: ' from esxi'
+    }
+  }
+
+  async getTransferableVmMetadata(vmId) {
+    const search = await this.search('VirtualMachine', ['name', 'config', 'storage', 'runtime', 'snapshot'])
+    if (search[vmId] === undefined) {
+      throw new Error(`VM ${vmId} not found `)
+    }
+    const { config, runtime } = search[vmId]
+
+    const [, dataStore, vmxPath] = config.files.vmPathName.match(/^\[(.*)\] (.+.vmx)$/)
+    const res = await this.download(dataStore, vmxPath)
+
+    const vmx = parseVmx(await res.text())
+
+    // list datastores
+    const dataStores = {}
+    Object.values(await this.search('Datastore', ['summary'])).forEach(({ summary }) => {
+      dataStores[summary.url] = summary
+    })
+
+    const disks = []
+
+    for (let scsiIndex = 0; scsiIndex < MAX_SCSI; scsiIndex++) {
+      const scsiChannel = vmx[`scsi${scsiIndex}`]
```

Comment:
See [`lodash.pickBy`](https://devdocs.io/lodash~4/index#pickBy).

Reply:
```js
for (const key of Object.keys(vmx)) {
  if (!key.startsWith('scsi')) {
    continue
  }
```

Reply:
This makes sure we handle all the necessary keys, even if there are more than 10.